### ISSUE TITLE
smarter BelongsTo relationship 

### ISSUE THREAD

**Issue by if4lcon** [laravel/ideas#1731]
if you have hasMany relationship and you try to get that relationship with belongsTo instance 
.. eloquent will fire new query and get relation example in code :

```php

$user = User::first();
$post = $user->posts()->first();
$post->user //new query with or without eager loading
```

I think belongsTo  should look for $user before making mysql query .

**Comment by ollieread** [comment]
That would be an entity manager. It would require that Eloquent keeps track of all models that it hydrates and references those. It would also need to keep track of when a model was updated in the database so that it didn't serve old data.

It's doable, but bigger than a `belongsTo` relationship.

**Comment by Patryk27** [comment]
This technique is called `identity map` and it's used extensively in Doctrine - implementing it in Eloquent wouldn't be probably that hard, but could change observable behavior.

```php
$user = User::first();
$user->foo = 'bar';

$post = $user->posts()->first();
dump($post->user->foo); // should return 'bar' or `null`?
```

**Comment by if4lcon** [comment]
🤔 should return bar I think but how it’s simple to implement

**Comment by mfn** [comment]
Doctrine already works that way. Realistically not going to happen for Eloquent; however everyone is free to use Doctrine also in Laravel, no one is forced to use Eloquent.

**Comment by if4lcon** [comment]
@mfn it's just an idea for future Laravel versions . it's hard to implement right now

**Comment by abdumu** [comment]
@mfn why the defensive tone! he suggested an idea in Laravel ideas hub. 

Also it is not true that it's not on the table for Laravel, as Tylor himself said once it's an INTERESTING  CONCEPT but he didn't find the right implementation.
(see: https://github.com/laravel/framework/pull/22401) 


Thanks.

**Comment by mfn** [comment]
I understand; well, I read all ideas posted and just gave a realistic impression.

Mark your calendar in 3 years and check back if it changed 🤷‍♀️

**Comment by if4lcon** [comment]
It's not big deal and not really important issue right now but it's good to have it at some point any performance improvements are welcome

**Comment by if4lcon** [comment]
Yeah and I would like to see it in Eloquent because we all use it :)
